lib/dynflow/actor.rb

```diff
@@ -7,10 +7,82 @@ def on_message(message)
     end
   end
 
+  # Extend the Concurrent::Actor::Envelope to include information about the origin of the message
+  module EnvelopeBacktraceExtension
+    def initialize(*args)
+      super
+      @origin_backtrace = caller + Actor::BacktraceCollector.current_actor_backtrace
+    end
+
+    def origin_backtrace
+      @origin_backtrace
+    end
+
+    def inspect
+      "#<#{self.class.name}:#{object_id}> @message=#{@message.inspect}, @sender=#{@sender.inspect}, @address=#{@address.inspect}>"
+    end
+  end
+  Concurrent::Actor::Envelope.prepend(EnvelopeBacktraceExtension)
+
   # Common parent for all the Dynflow actors defining some defaults
   # that we preffer here.
   class Actor < Concurrent::Actor::Context
+    module LogWithFullBacktrace
+      def log(level, message = nil, &block)
+        if message.is_a? Exception
+          error = message
+          backtrace = Actor::BacktraceCollector.full_backtrace(error.backtrace)
+          log(level, format("%s (%s)\n%s", error.message, error.class, backtrace.join("\n")))
+        else
+          super
+        end
+      end
+    end
+
+    class SetResultsWithOriginLogging < Concurrent::Actor::Behaviour::SetResults
+      include LogWithFullBacktrace
+
+      def on_envelope(envelope)
+        Actor::BacktraceCollector.with_backtrace(envelope.origin_backtrace) { super }
+      end
+    end
+
+    class BacktraceCollector
+      CONCURRENT_RUBY_LINE = '[ concurrent-ruby ]'.freeze
+
+      class << self
+        def with_backtrace(backtrace)
+          previous_actor_backtrace = Thread.current[:_dynflow_actor_backtrace]
+          Thread.current[:_dynflow_actor_backtrace] = backtrace
+          yield
+        ensure
+          Thread.current[:_dynflow_actor_backtrace] = previous_actor_backtrace
+        end
+
+        def current_actor_backtrace
+          Thread.current[:_dynflow_actor_backtrace] || []
+        end
+
+        def full_backtrace(backtrace)
+          filter_backtrace((backtrace || []) + current_actor_backtrace)
+        end
+
+        private
+
+        def filter_line?(line)
+          %w[concurrent-ruby gems/logging actor.rb].any? { |pattern| line.include?(pattern) }
```

Comment:
Wouldn't this also match lines coming from `lib/dynflow/actor.rb`?

Reply:
It's on purpose, as I consider that still as part of plumbing

Reply:
Ok, nevermind then.